util/linuxfw/iptable_runner.go

```diff
@@ -0,0 +1,506 @@
+				// the desired state anyway.
+				return nil
+			}
+			return fmt.Errorf("flushing %s/%s: %w", table, chain, err)
+		}
+		return nil
+	}
+
+	for _, ipt := range i.getTables() {
+		if err := del(ipt, "filter", "ts-input"); err != nil {
+			return err
+		}
+		if err := del(ipt, "filter", "ts-forward"); err != nil {
+			return err
+		}
+	}
+	if err := del(i.ipt4, "nat", "ts-postrouting"); err != nil {
+		return err
+	}
+	if i.v6NATAvailable {
+		if err := del(i.ipt6, "nat", "ts-postrouting"); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// delNetfilterHooks deletes the calls to tailscale's netfilter chains
+// in the relevant main netfilter chains.
+func (i *iptablesRunner) DelNetfilterHooks(logf logger.Logf) error {
+	del := func(ipt iptablesInterface, table, chain string) error {
+		tsChain := tsChain(chain)
+		args := []string{"-j", tsChain}
+		if err := ipt.Delete(table, chain, args...); err != nil {
+			// TODO(apenwarr): check for errCode(1) here.
+			// Unfortunately the error code from the iptables
+			// module resists unwrapping, unlike with other
+			// calls. So we have to assume if Delete fails,
+			// it's because there is no such rule.
+			logf("note: deleting %v in %s/%s: %w", args, table, chain, err)
+			return nil
+		}
+		return nil
+	}
+
+	for _, ipt := range i.getTables() {
+		if err := del(ipt, "filter", "INPUT"); err != nil {
+			return err
+		}
+		if err := del(ipt, "filter", "FORWARD"); err != nil {
+			return err
+		}
+	}
+	if err := del(i.ipt4, "nat", "POSTROUTING"); err != nil {
+		return err
+	}
+	if i.v6NATAvailable {
+		if err := del(i.ipt6, "nat", "POSTROUTING"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// addSNATRule adds a netfilter rule to SNAT traffic destined for
+// local subnets.
+func (i *iptablesRunner) AddSNATRule() error {
+
+	args := []string{"-m", "mark", "--mark", TailscaleSubnetRouteMark + "/" + TailscaleFwmarkMask, "-j", "MASQUERADE"}
+	if err := i.ipt4.Append("nat", "ts-postrouting", args...); err != nil {
+		return fmt.Errorf("adding %v in v4/nat/ts-postrouting: %w", args, err)
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Append("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("adding %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// delSNATRule removes the netfilter rule to SNAT traffic destined for
+// local subnets. Fails if the rule does not exist.
+func (i *iptablesRunner) DelSNATRule() error {
+
+	args := []string{"-m", "mark", "--mark", TailscaleSubnetRouteMark + "/" + TailscaleFwmarkMask, "-j", "MASQUERADE"}
+	if err := i.ipt4.Delete("nat", "ts-postrouting", args...); err != nil {
+		return fmt.Errorf("deleting %v in v4/nat/ts-postrouting: %w", args, err)
+	}
+	if i.v6NATAvailable {
+		if err := i.ipt6.Delete("nat", "ts-postrouting", args...); err != nil {
+			return fmt.Errorf("deleting %v in v6/nat/ts-postrouting: %w", args, err)
+		}
+	}
+	return nil
+}
+
+// Cleans the iptables rules created by Tailscale.
```

Comment:
We should start docs with the function name, so in this case

```suggestion
// IptablesCleanUp clears the iptables rules for IPv4 and IPv6. If an error occurs,
// it is logged.
```

Reply:
With this change, I think you could also remove the 

```
// Clear iptables rules
and
// Clear ip6tables rules
and 
// If the system doesn't support ipv6, we'll just log the error and continue
```

comments